File: persephone/config-service/src/jvm/com/vmware/blockchain/deployment/service/server/ConfigurationServiceModule.java
/*
 * Copyright (c) 2019 VMware, Inc. All rights reserved. VMware Confidential
 */

package com.vmware.blockchain.deployment.service.server;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.ForkJoinPool;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * {@link Singleton}-scoped Dagger {@link Module} that declares the components and dependencies
 * for {@link ConfigurationService} instance.
 */
@Module
public class ConfigurationServiceModule {
    /**
     * Provide an {@link ExecutorService} instance.
     *
     * @return
     *   a singleton {@link ExecutorService} instance.
     */
    @Provides
    @Singleton
    static ExecutorService providesExecutorService() {
        return ForkJoinPool.commonPool();
    }

    /**
     * Provide an {@link ConfigurationService} instance.
     *
     * @return
     *   a singleton {@link ConfigurationService} instance.
     */
    @Provides
    @Singleton
    static ConfigurationService providesConfigurationService(ExecutorService executor) {
        return new ConfigurationService(executor);
    }
}

File: persephone/provisioning-service/src/jvm/com/vmware/blockchain/deployment/service/provision/ConfigurationServiceModule.kt
/* **************************************************************************
 * Copyright (c) 2019 VMware, Inc. All rights reserved. VMware Confidential
 * **************************************************************************/
package com.vmware.blockchain.deployment.service.provision

import com.vmware.blockchain.deployment.model.ConfigurationServiceStub
import com.vmware.blockchain.deployment.model.Endpoint
import com.vmware.blockchain.deployment.orchestration.vmware.newClientRpcChannel
import dagger.Module
import dagger.Provides
import io.grpc.CallOptions
import javax.inject.Singleton

@Module
class ConfigurationServiceModule {

    /**
     * Provide an [ConfigurationServiceStub] instance.
     *
     * @param[configurationService]
     *   endpoint specification for configuration service.
     *
     * @return
     *   a singleton [ConfigurationServiceStub] instance
     */
    @Provides
    @Singleton
    fun providesConfigurationServiceStub(configurationService: Endpoint): ConfigurationServiceStub {
        // Note: Endpoint.newClientRpcChannel() is currently exported from orchestration-vmware
        // library module. This will need to be moved.
        val channel = configurationService.newClientRpcChannel()
        return ConfigurationServiceStub(channel, CallOptions.DEFAULT)
    }
}